docs: #1561 Tiered Storage fine grained caching (#1596)
redpanda-data/documentation: modules/manage/pages/cluster-maintenance/about-throughput-quotas.adoc

Use client-based quotas when:

* Authentication is not enabled.
* Grouping by application or service name is sufficient.
* You operate a single-tenant environment where all clients are trusted.
* You need simple rate limiting without user-level isolation.

=== User-based quotas

IMPORTANT: User-based quotas require xref:manage:security/authentication.adoc[authentication] to be enabled on your cluster.

User-based quotas apply to authenticated user principals (from SASL, mTLS, or OIDC authentication). Each user has a separate quota, providing true multi-tenant isolation.

User-based quotas rely on Redpanda's authentication system to verify user identity. The user principal is extracted from SASL credentials, mTLS certificates, or OIDC tokens and cannot be forged by clients. This provides true security isolation between tenants.

Use user-based quotas when:

* You operate a multi-tenant environment, such as SaaS platforms or enterprises with departments.
* You require security isolation between users or tenants, including for compliance requirements.

> * You need per-user billing or metering.

Reviewer: wording: I think I would not use the "security isolation" and "multi-tenant isolation" wording here because that could imply additional guarantees about how the requests from different users/clients are processed.
```suggestion
User-based quotas apply to authenticated user principals (from SASL, mTLS, or OIDC authentication). Each user has a separate quota, providing a way to limit the impact of individual users on the cluster.

User-based quotas rely on Redpanda's authentication system to verify user identity. The user principal is extracted from SASL credentials, mTLS certificates, or OIDC tokens and cannot be forged by clients.

Use user-based quotas when:

* You operate a multi-tenant environment, such as SaaS platforms or enterprises with departments.
* You require isolation between users or tenants, including for compliance requirements.
* You need per-user billing or metering.
```

rpk cluster quotas alter --add consumer_byte_rate=20000000 --name client-id=app-1
----

|===
| User + Client ID | Precedence match | Resulting quota

| `user="alice"`, `client_id="app-1"`
| Level 1: Exact user + exact client
| 5 MB/s

| `user="alice"`, `client_id="app-2"`
| Level 4: Exact user only
| 10 MB/s

| `user="bob"`, `client_id="app-1"`
| Level 9: Exact client only
| 20 MB/s

| `user="bob"`, `client_id="app-2"`
| Level 12: No quota configured
| Unlimited
|===

=== Unauthenticated connections

Unauthenticated connections have an empty user principal (`user=""`), which is treated as a specific entity, not as `user=<default>`.

Unauthenticated connections:

* Fall back to client-only quotas (precedence levels 9-11).
* Have unlimited throughput only if no client-only quota matches.

== Quota tracking and isolation

Each unique (user principal + `client_id`) combination is tracked independently for quota enforcement, even if multiple combinations have the same quota limit configured. This means:

* Different users using the same `client_id` each get their own quota tracking.
* The same user with different `client_id` values also gets separate quota tracking with each client.

> * Quota usage is not shared between different (user, client) pairs.

Reviewer: This is a very useful section, probably the thing I end up explaining most often :) However, unfortunately the behaviour is not this simple.

Claude and I came up with this suggestion:

---

Redpanda tracks quota usage using a tracker key that determines which connections share the same quota bucket. The tracker key—and therefore the isolation level—depends on which quota rule matches the request, following a priority-based hierarchy.

 ### Tracking granularity by quota type
```
| Quota configuration level                 | Tracker key       | Isolation behavior                                                       |
|-------------------------------------------|-------------------|--------------------------------------------------------------------------|
| users/<user>/clients/<client-id>          | (user, client_id) | Each unique (user, client_id) pair tracked independently                 |
| users/<user>/client-id-prefix/<prefix>    | (user, prefix)    | Clients matching the prefix share tracking within that user              |
| users/<user>/clients/<default>            | (user, client_id) | Each unique (user, client_id) pair tracked independently                 |
| users/<user>                              | user              | All clients for that user share a single tracking bucket                 |
| users/<default>/clients/<client-id>       | (user, client_id) | Each unique (user, client_id) pair tracked independently                 |
| users/<default>/client-id-prefix/<prefix> | (user, prefix)    | Clients matching the prefix share tracking within each user              |
| users/<default>/clients/<default>         | (user, client_id) | Each unique (user, client_id) pair tracked independently                 |
| users/<default>                           | user              | All clients for each user share a single tracking bucket (per user)      |
| clients/<client-id>                       | client_id         | All users with that client_id share a single tracking bucket             |
| client-id-prefix/<prefix>                 | prefix            | All clients matching the prefix share a single bucket (across all users) |
| clients/<default>                         | client_id         | Each unique client_id tracked independently                              |
| No quota configured                       | N/A               | No tracking / unlimited throughput                                       |
```
###  Key behaviors

  - User-only quotas share across client_ids: When a quota is configured at the user level (e.g., user/alice) without specifying a client_id dimension, all connections from that user share a single quota bucket regardless of their client_id.
  - Client-id-only quotas share across users: When a quota is configured at the client_id level (e.g., client-id/my-app) without specifying a user dimension, all users connecting with that client_id share a single quota bucket.
  - To get independent quota tracking per (user, client_id) combination, you must configure quotas that include both dimensions, such as user/<user>/client-id/<client-id> or
  user/<default>/client-id/<default>.
  - Prefix-based quotas group clients: Client ID prefix quotas (e.g., client-id-prefix/app-) cause all clients whose client_id starts with that prefix to share a quota bucket.

### Example: User-only quota

  If you configure:
  /config/users/alice → producer_byte_rate: 10MB/s

  Then Alice connecting with client_id=app1 and Alice connecting with client_id=app2 share the same 10 MB/s limit. Their combined throughput cannot exceed 10 MB/s.

  To give each of Alice's clients an independent 10 MB/s limit, configure:
  /config/users/alice/clients/<default> → producer_byte_rate: 10MB/s

### Example: User default quota

  If you configure:
  /config/users/<default> → producer_byte_rate: 10MB/s

  This quota applies to all users who don't have a more specific quota configured. However, each user is tracked independently—Alice gets her own 10 MB/s bucket, Bob gets his own 10 MB/s bucket, and so on.

  Within each user, all client_id values share that user's bucket. So Alice connecting with client_id=app1 and Alice connecting with client_id=app2 share the same 10 MB/s limit, while Bob's connections (regardless of his client_id) have a separate
  10 MB/s limit.

Reviewer 2: Added

:page-topic-type: concepts
:page-aliases:
:personas: platform_admin, developer
:learning-objective-1: Describe the difference between user-based and client-based quotas
:learning-objective-2: Determine which quota type to use for your use case
:learning-objective-3: Explain quota precedence rules and how Redpanda tracks quota usage

// tag::single-source[]
ifdef::env-cloud[]
:authentication-doc: security:cloud-authentication.adoc
endif::[]
ifndef::env-cloud[]
:authentication-doc: manage:security/authentication.adoc
endif::[]

Redpanda uses throughput quotas to limit the rate of produce and consume requests from clients. Understanding how quotas work helps you prevent individual clients from disproportionately consuming resources and causing performance degradation for other clients (also known as the "noisy-neighbor" problem), and ensure fair resource sharing across users and applications.

After reading this page, you will be able to:

* [ ] {learning-objective-1}
* [ ] {learning-objective-2}
* [ ] {learning-objective-3}

To configure and manage throughput quotas, see xref:manage:cluster-maintenance/manage-throughput.adoc[].

== Throughput control overview

> Redpanda provides two types of throughput limits: broker-wide limits configured using cluster properties and client quotas configured using Kafka API. 

Reviewer 3: Maybe format these two as definition lists? Easier to consume quickly.

Reviewer 2: Updated to use unordered lists

> This page covers client quotas, which enable per-user and per-client rate limiting with fine-grained control through entity hierarchy and precedence rules. For information about broker-wide limits, see xref:manage:cluster-maintenance/manage-throughput.adoc#broker-wide-throughput-limits[Broker-wide throughput limits].

Reviewer 3: This confused me: "This page"...we don't typically use that to describe the content, do we?  I see that client quotas is a sub-topic within Quota entities, but there are also other topics covered here too, like user-based quotas. 

Is there an issue here with heading hierarchy?

Reviewer 2: We have used "This page" in other docs. Updated to use "Client ID-based" when referring specifically to quota rules that use the client ID (or ID prefix) for matching

== Supported quota types

Redpanda supports three Kafka API-based quota types:

|===
| Quota type | Description

| `producer_byte_rate`
| Limit throughput of produce requests (bytes per second)

| `consumer_byte_rate`
| Limit throughput of fetch requests (bytes per second)

| `controller_mutation_rate`
| Limit rate of topic mutation requests (partitions created or deleted per second)
|===

All quota types can be applied to user-based entities, client-based entities, or combinations of both.

== Quota entities

Redpanda supports two primary entity types for client quotas:

> * Client-based quotas (`client_id` or `client_id` prefix): Limit throughput based on the `client_id` field that clients self-declare. Multiple clients using the same `client_id` share the quota. This quota does not require authentication.

Reviewer 4: "Does not require users to be authenticated"

Reviewer 2: Updated

* User-based quotas: Limit throughput based on user glossterm:principal[] authenticated through SASL, mTLS, or OIDC.

You can also combine both types for fine-grained control, such as limiting a specific user when using a specific client application.

=== Client-based quotas

Client-based quotas apply to clients identified by their `client_id` field, which is set by the client application. The `client_id` field is typically a configurable property when you create a client with Kafka libraries. Multiple clients using the same `client_id` share the same quota tracking.

> Client-based quotas rely on clients honestly reporting their identity. Because `client_id` is self-declared, any client can claim any identity. This makes client-based quotas unsuitable for security isolation between tenants.

Reviewer 4: "This makes client quotas rely on the 'honor system' and the accurate setting of client_id property and tis thus unsuitable for guaranteeing isolation between tenants"

(not really about security)

Reviewer 2: Updated and removed "security"

Use client-based quotas when:

* Authentication is not enabled.
* Grouping by application or service name is sufficient.
* You operate a single-tenant environment where all clients are trusted.
* You need simple rate limiting without user-level isolation.

=== User-based quotas

IMPORTANT: User-based quotas require xref:manage:security/authentication.adoc[authentication] to be enabled on your cluster.

User-based quotas apply to authenticated user principals. Each user has a separate quota, providing a way to limit the impact of individual users on the cluster.

User-based quotas rely on Redpanda's authentication system to verify user identity. The user principal is extracted from SASL credentials, mTLS certificates, or OIDC tokens and cannot be forged by clients.

Use user-based quotas when:

* You operate a multi-tenant environment, such as SaaS platforms or enterprises with departments.

> * You require isolation between users or tenants, including for compliance requirements.

Reviewer 4: "* You require isolation between users or tenants, to avoid noisy neighbor issues."

Reviewer 2: @mattschumpert Updated this line, and also updated the intro to this doc: https://deploy-preview-1596--redpanda-docs-preview.netlify.app/current/manage/cluster-maintenance/about-throughput-quotas/

> Redpanda uses throughput quotas to limit the rate of produce and consume requests from clients. Understanding how quotas work helps you prevent individual clients from disproportionately consuming resources and causing performance degradation for other clients (also known as the "noisy-neighbor" problem), and ensure fair resource sharing across users and applications.

How does this sound?

* You need per-user billing or metering.

=== Combined user and client quotas

You can combine user and client identities for fine-grained control over specific (user, client) combinations.

Use combined quotas when:

* You need fine-grained control, for example: "user `alice` using specific application".

> * Different rate limits apply to different apps used by the same user, for example: "`alice`'s payment-processor gets 10 MB/s, but `alice`'s analytics-consumer gets 50 MB/s". See <<Quota tracking and isolation>> for examples.

Reviewer 3: ```suggestion
* Different rate limits apply to different apps used by the same user. For example: "`alice`'s payment-processor gets 10 MB/s, but `alice`'s analytics-consumer gets 50 MB/s". See <<Quota tracking and isolation>> for examples.
```

Reviewer 3: Not sure I understand the need/use of the double quotation marks here?

| 9
| `/config/clients/<client-id>`
| `client_id`
| All users with that client_id share a single tracking bucket

| 10
| `/config/client-id-prefix/<prefix>`
| `prefix`
| All clients matching the prefix share a single bucket (across all users)

| 11
| `/config/clients/<default>`
| `client_id`
| Each unique client_id tracked independently

| 12
| No quota configured
| N/A
| No tracking / unlimited throughput
|===

* User-only quotas share across client_ids: When you configure a quota at the user level (for example, `/config/users/alice`) without specifying a `client_id` dimension, all connections from that user share a single quota bucket regardless of their `client_id`.
* Client-id-only quotas share across users: When you configure a quota at the `client_id` level (for example, `/config/clients/app-1`) without specifying a user dimension, all users connecting with that `client_id` share a single quota bucket.
* Combined quotas for per-(user, client_id) tracking: To get independent quota tracking per (user, client_id) combination, configure quotas that include both dimensions, such as `/config/users/<user>/clients/<client-id>` or `/config/users/<default>/clients/<default>`.

> * Prefix-based quotas group clients: Client ID prefix quotas (for example, `/config/client-id-prefix/app-`) cause all clients whose `client_id` starts with that prefix to share a quota bucket.

Reviewer 3: Wondering if each of these bullets applies to a specific level shown above. If so, you could use numbers to help readers quickly make the match/connection. I could be all wrong here, but if not, it sure would help me to quickly make the connection between the bullets here and the various levels above.

Reviewer 2: These sections have been reorganized: https://deploy-preview-1596--redpanda-docs-preview.netlify.app/current/manage/cluster-maintenance/about-throughput-quotas/#quota-precedence-and-tracking

Only kept list item #3 as the others are explained in the table itself

=== Example: User-only quota

If you configure a 10 MB/s produce quota for user `alice`:

[,bash]
----
rpk cluster quotas alter --add producer_byte_rate=10000000 --name user=alice
----

Then `alice` connecting with `client_id=app-1` and `alice` connecting with `client_id=app-2` share the same 10 MB/s limit. Their combined throughput cannot exceed 10 MB/s.

> To give each of `alice`'s clients an independent 10 MB/s limit, configure:

Reviewer 3: ```suggestion
To give each of `alice`'s` clients an independent 10 MB/s limit, configure:
```

Reviewer 3: Was missing the closing delimiter, but is this what you intended?

Reviewer 2: This required a passthrough to so that the text renders correctly

[,bash]
----
rpk cluster quotas alter --add producer_byte_rate=10000000 --name user=alice --default client-id
----

=== Example: User default quota

If you configure a default 10 MB/s produce quota for all users:

[,bash]
----
rpk cluster quotas alter --add producer_byte_rate=10000000 --default user
----

> This quota applies to all users who don't have a more specific quota configured. Each user is tracked independently: `alice` gets their own 10 MB/s bucket, `bob` gets their own 10 MB/s bucket, and so on.

Reviewer 3: ```suggestion
This quota applies to all users who don't have a more specific quota configured. Each user is tracked independently: `alice` gets her own 10 MB/s bucket, `bob` gets his own 10 MB/s bucket, and so on.
```

> Within each user, all `client_id` values share that user's bucket. `alice` connecting with `client_id=app-1` and `alice` connecting with `client_id=app-2` share the same 10 MB/s limit, while `bob`'s connections have a separate 10 MB/s limit.

Reviewer 3: ```suggestion
Within each user, all `client_id` values share that user's bucket. `alice` connecting with `client_id=app-1` and `alice` connecting with `client_id=app-2` share the same 10 MB/s limit, while `bob`'s` connections have a separate 10 MB/s limit.
```

Reviewer 3: Again, not sure if this is what you intended--just adding a closing delimiter.

Reviewer 2: Added passthrough
